File: examples/openai-agents-sds-mcp-list/src/policy.mjs
import {
  ALLOWED_HTTP_METHODS,
  ALLOWED_RPC_METHODS,
  CREDENTIAL_HEADER_NAMES,
  LIVE_MCP_HOST,
  LIVE_MCP_PATH,
  REFUSED_CLI_FLAGS,
} from "./constants.mjs";
import { PolicyRefusal } from "./errors.mjs";

const ALLOWED_RPC = new Set(ALLOWED_RPC_METHODS);
const ALLOWED_HTTP = new Set(ALLOWED_HTTP_METHODS);
const CREDENTIAL_HEADERS = new Set(CREDENTIAL_HEADER_NAMES);

export function assertMcpListUrl(raw) {
  let url;
  try {
    url = new URL(String(raw || ""));
  } catch {
    throw new PolicyRefusal("url must be an absolute HTTP(S) URL", { field: "url" });
  }
  if (url.username || url.password) {
    throw new PolicyRefusal("url must not contain credentials", { field: "url" });
  }
  if (url.hash) {
    throw new PolicyRefusal("url must not contain a fragment", { field: "url" });
  }
  if (url.search) {
    throw new PolicyRefusal("url must not contain a query", { field: "url" });
  }
  if (url.pathname !== LIVE_MCP_PATH) {
    throw new PolicyRefusal("url path must be /mcp", { field: "url" });
  }

  const host = url.hostname.toLowerCase();
  if (host === LIVE_MCP_HOST) {
    if (url.protocol !== "https:") {
      throw new PolicyRefusal("SameDayDesk MCP url must use HTTPS", { field: "url" });
    }
    if (url.port && url.port !== "443") {
      throw new PolicyRefusal("SameDayDesk MCP url must use the default HTTPS port", { field: "url" });
    }
    return `${url.protocol}//${url.hostname}${url.pathname}`;
  }

  if (host === "127.0.0.1" || host === "localhost") {
    if (url.protocol !== "http:" && url.protocol !== "https:") {
      throw new PolicyRefusal("loopback MCP url must use HTTP(S)", { field: "url" });
    }
    return url.origin + url.pathname;
  }

  throw new PolicyRefusal("url host is not SameDayDesk MCP or loopback", { field: "url" });
}

export function assertNoCredentialHeaders(headers) {
  const names = headerNames(headers);
  for (const name of names) {
    if (CREDENTIAL_HEADERS.has(name)) {
      throw new PolicyRefusal(`credential or payment header refused: ${name}`, { field: name });
    }
  }
}

export function assertAllowedHttpMethod(method) {
  const normalized = String(method || "GET").toUpperCase();
  if (!ALLOWED_HTTP.has(normalized)) {
    throw new PolicyRefusal(`HTTP method ${normalized} is refused`, { field: "method" });
  }
  return normalized;
}

export function assertAllowedRpcMethod(method) {
  if (method == null || method === "") return null;
  const name = String(method);
  if (name === "tools/call") {
    throw new PolicyRefusal("tools/call is refused; this example only lists tools unpaid", {
      field: "method",
    });
  }
  if (!ALLOWED_RPC.has(name)) {
    throw new PolicyRefusal(
      `MCP method ${name} is refused; this example only lists tools unpaid`,
      { field: "method" },
    );
  }
  return name;
}

export function refusedCliFlag(arg) {
  const token = String(arg || "");
  const flag = token.includes("=") ? token.slice(0, token.indexOf("=")) : token;
  if (REFUSED_CLI_FLAGS.includes(flag)) return flag;
  return null;
}

function headerNames(headers) {
  if (!headers) return [];
  if (typeof headers.forEach === "function") {
    const names = [];
    headers.forEach((_value, name) => names.push(String(name).toLowerCase()));
    return names;
  }
  if (Array.isArray(headers)) {
    return headers.map((entry) => String(entry?.[0] || "").toLowerCase());
  }
  return Object.keys(headers).map((name) => name.toLowerCase());
}

File: examples/crewai-sds-mcp-list/src/crewai_sds_mcp_list/policy.py
"""Local unpaid-discovery policy. Runs before any MCP transport is opened."""

from __future__ import annotations

from ipaddress import ip_address
from urllib.parse import urlparse

from .constants import LIVE_HOST, LIVE_MCP_PATH, LOOPBACK_HOST
from .errors import PaymentRefused, PolicyError, ToolsCallRefused

FORBIDDEN_HEADER_NAMES = frozenset(
    {
        "authorization",
        "proxy-authorization",
        "cookie",
        "set-cookie",
        "payment-signature",
        "x-payment",
        "x-payment-signature",
        "x402-payment",
        "x402/payment",
        "mcp-method",
        "www-authenticate",
        "x-api-key",
        "api-key",
    }
)

FORBIDDEN_HEADER_NEEDLES = (
    "bearer ",
    "api_key",
    "api-key",
    "private_key",
    "private-key",
    "x402/payment",
    "payment-signature",
)

FORBIDDEN_CLI_FLAGS = frozenset(
    {
        "--approve",
        "--call",
        "--pay",
        "--wallet",
        "--private-key",
        "--private-key-env",
        "--payment-signature",
        "--authorization",
    }
)


def tool_name(tool: dict) -> str:
    original = tool.get("original_name")
    if isinstance(original, str) and original:
        return original
    name = tool.get("name")
    if isinstance(name, str) and name:
        return name
    return ""


def assert_no_tools_call(*, attempted: bool = False, tool: str | None = None) -> None:
    if attempted or tool:
        suffix = f" ({tool})" if tool else ""
        raise ToolsCallRefused(
            "this example never issues MCP tools/call"
            f"{suffix}; unpaid tools/list only"
        )


def assert_no_forbidden_cli_flags(argv: list[str]) -> None:
    for arg in argv:
        name = arg.split("=", 1)[0]
        if name in FORBIDDEN_CLI_FLAGS:
            if name in {"--call"}:
                raise ToolsCallRefused(
                    f"{name} is refused: this example never issues MCP tools/call"
                )
            raise PaymentRefused(
                f"{name} is refused: unpaid tools/list sends no wallet, "
                "signature, or payment header"
            )


def _header_items(headers: dict | None) -> list[tuple[str, str]]:
    if headers is None:
        return []
    if not isinstance(headers, dict) or isinstance(headers, list):
        raise PolicyError("MCPServerHTTP.headers must be a dict of strings or None")
    items = []
    for key, value in headers.items():
        if not isinstance(key, str) or not isinstance(value, str):
            raise PolicyError("MCPServerHTTP.headers keys and values must be strings")
        items.append((key, value))
    return items


def assert_unpaid_headers(headers: dict | None) -> dict[str, str]:
    items = _header_items(headers)
    for key, value in items:
        lower = key.lower().strip()
        if lower in FORBIDDEN_HEADER_NAMES:
            raise PaymentRefused(
                f"refusing header {key!r}: unpaid tools/list sends no "
                "credentials, payment, or Mcp-Method"
            )
        blob = f"{key}:{value}".lower()
        if any(needle in blob for needle in FORBIDDEN_HEADER_NEEDLES):
            raise PaymentRefused(
                f"refusing header {key!r}: value looks like a credential or payment"
            )
    return {key: value for key, value in items}


def assert_discovery_url(raw: str, *, allow_loopback: bool = False) -> str:
    if not isinstance(raw, str) or not raw.strip():
        raise PolicyError("MCP URL is required")
    try:
        parsed = urlparse(raw)
    except Exception as exc:
        raise PolicyError("MCP URL is not a valid URL") from exc
    if parsed.username or parsed.password:
        raise PolicyError("URL credentials are refused")
    if parsed.query or parsed.fragment:
        raise PolicyError("MCP URL must not include query or fragment")
    if parsed.path != LIVE_MCP_PATH:
        raise PolicyError(f"MCP path must be exactly {LIVE_MCP_PATH}")
    host = (parsed.hostname or "").lower()
    if not host:
        raise PolicyError("MCP URL host is required")
    if ":" in host:
        raise PolicyError("IPv6 MCP URLs are refused")

    if host == LIVE_HOST:
        if parsed.scheme != "https":
            raise PolicyError("live SameDayDesk MCP is HTTPS only")
        if parsed.port not in (None, 443):
            raise PolicyError("live SameDayDesk MCP uses port 443")
        return raw

    if allow_loopback and host == LOOPBACK_HOST:
        if parsed.scheme != "http":
            raise PolicyError("loopback fixture MCP is HTTP on 127.0.0.1 only")
        if parsed.port is None or not (1 <= parsed.port <= 65535):
            raise PolicyError("loopback fixture MCP requires an explicit port")
        try:
            if ip_address(host).is_loopback is False:
                raise PolicyError("loopback fixture host is not loopback")
        except ValueError as exc:
            raise PolicyError("loopback fixture host is not a valid IP") from exc
        return raw

    raise PolicyError(
        "MCP URL must be the live SameDayDesk origin "
        "https://agents.samedaydesk.com/mcp "
        "(loopback 127.0.0.1 is allowed only for the seeded fixture)"
    )


def assert_unpaid_mcp_server_http(config, *, allow_loopback: bool = False) -> None:
    # Imported lazily so policy unit tests that only check URLs stay cheap, and
    # so CREWAI_DISABLE_TELEMETRY is set by callers before crewai loads.
    from crewai.mcp import MCPServerHTTP

    if type(config) is not MCPServerHTTP and not isinstance(config, MCPServerHTTP):
        raise PolicyError("config must be crewai.mcp.MCPServerHTTP")
    assert_discovery_url(config.url, allow_loopback=allow_loopback)
    assert_unpaid_headers(config.headers)
    if config.streamable is not True:
        raise PolicyError(
            "SameDayDesk remote MCP is streamable HTTP; MCPServerHTTP.streamable must be True"
        )
    if config.tool_filter is not None:
        raise PolicyError(
            "tool_filter is refused: unpaid discovery must not hide extract or extract_batch"
        )
